src/language/rule/rule.ts

```diff
@@ -265,3 +278,37 @@ export class RuleFailure {
         return new RuleFailurePosition(position, lineAndCharacter);
     }
 }
+
+export class WalkContext<T> {
+    private limit: number;
+    private failures: RuleFailure[];
+
+    constructor(public readonly sourceFile: ts.SourceFile, public readonly ruleName: string, public readonly options: T) {
+        this.failures = [];
+        this.limit = sourceFile.getFullWidth();
+    }
+
+    public getFailures() {
+        return this.failures;
+    }
+
+    /** Add a failure with any arbitrary span. Prefer `addFailureAtNode` if possible. */
+    public addFailureAt(start: number, width: number, failure: string, fix?: Fix) {
+        this.addFailure(start, start + width, failure, fix);
+    }
+
+    public addFailure(start: number, end: number, failure: string, fix?: Fix) {
```

Comment:
Sure, we should try to migrate all existing rules, even `ScopeAwareRuleWalker` and subclasses thereof. At least that's my naive plan.

I don't know if or when we can eventually deprecate and remove `RuleWalker` for implementers of custom rules. In the end it's your decision if you want to keep it for backwards compatibility.

Regarding the method name I don't have a strong opinion. I will change it if you want.